savant/converter/yolo.py

```diff
@@ -56,92 +104,106 @@ def __call__(
-
-        # scale
-        if model.input.maintain_aspect_ratio:
-            bboxes /= min(
-                model.input.width / roi_width,
-                model.input.height / roi_height,
+            mask = np.isin(class_ids, self.class_ids)
+            bboxes, confidences, class_ids = (
+                bboxes[mask],
+                confidences[mask],
+                class_ids[mask],
+            )
+
+        # --- confidence filter
+        if self.confidence_threshold > 0:
+            mask = confidences > self.confidence_threshold
+            bboxes, confidences, class_ids = (
+                bboxes[mask],
+                confidences[mask],
+                class_ids[mask],
+            )
+
+        # --- NMS (class-agnostic)
+        if self.nms_iou_threshold > 0 and bboxes.shape[0] > 1:
+            keep = nms_cpu(bboxes, confidences, self.nms_iou_threshold, self.top_k)
+            bboxes, confidences, class_ids = (
+                bboxes[keep],
+                confidences[keep],
+                class_ids[keep],
+            )
+        elif bboxes.shape[0] > self.top_k:
+            idx = np.argpartition(confidences, -self.top_k)[-self.top_k :]
+            bboxes, confidences, class_ids = (
+                bboxes[idx],
+                confidences[idx],
+                class_ids[idx],
             )
+
+        # --- map from model input space -> ROI space
+        roi_left, roi_top, roi_w, roi_h = roi
+
+        if model.input.maintain_aspect_ratio:
+            # reverse letterbox: remove padding, then de-scale
+            scale = min(model.input.width / roi_w, model.input.height / roi_h)
+            if scale <= 0:
+                return None
+            # de-scale w,h,xc,yc
+            bboxes /= scale
+
+            if getattr(model.input, 'symmetric_padding', False):
+                new_w = roi_w * scale
+                new_h = roi_h * scale
+                pad_x = (model.input.width - new_w) / 2.0
+                pad_y = (model.input.height - new_h) / 2.0
+                # convert pad from model space to ROI space
+                pad_x /= scale
+                pad_y /= scale
+                bboxes[:, 0] -= pad_x  # xc
+                bboxes[:, 1] -= pad_y  # yc
         else:
-            bboxes[:, [0, 2]] /= model.input.width / roi_width
-            bboxes[:, [1, 3]] /= model.input.height / roi_height
+            # simple affine scaling
+            bboxes[:, [0, 2]] /= model.input.width / roi_w
+            bboxes[:, [1, 3]] /= model.input.height / roi_h
 
-        # correct xc, yc
+        # offset by ROI top-left
         bboxes[:, 0] += roi_left
         bboxes[:, 1] += roi_top
 
+        # --- pack result
         return np.concatenate(
             (
                 class_ids.reshape(-1, 1).astype(np.float32),
-                confidences.reshape(-1, 1),
-                bboxes,
+                confidences.reshape(-1, 1).astype(np.float32),
```

Comment:
Are you sure we need these type conversions (`astype`)? You specify the type at the very beginning, and it should not change during the ops.